Merge from onprc19.1 & onprc19.1Prod r.63271 to 65949

labkey-bpatel opened the pull request:
#### Rationale
Merge changes from onprc19.1 r.63271 to 65757
Merge changes from onprc19.1Prod r. 65762 to 65949

#### Related Pull Requests
https://github.com/LabKey/onprcEHRModules/pull/32
https://github.com/LabKey/platform/pull/1626
https://github.com/LabKey/DiscvrLabKeyModules/pull/64
https://github.com/LabKey/LabDevKitModules/pull/55
https://github.com/LabKey/sampleManagement/pull/381
https://github.com/LabKey/biologics/pull/712
https://github.com/LabKey/labbook/pull/53
https://github.com/LabKey/wnprc-modules/pull/38

labkey-bpatel committed:
Merge from onprc19.1 r.63271

labkey-bpatel committed:
Merge from onprc19.1 r.63282

labkey-bpatel committed:
Merge from onprc19.1 r.63282

labkey-bpatel committed:
Merge from onprc19.1 r.63328 to 64381

labkey-bpatel committed:
Merge from onprc19.1 r.64647

labkey-bpatel committed:
Merge from onprc19.1 r.64648

labkey-bpatel committed:
Merge from onprc19.1 r.64649 to 64652

labkey-bpatel committed:
Merge from onprc19.1 r.64748

labkey-bpatel committed:
Merge from onprc19.1 r.64953 to 65052

labkey-bpatel committed:
Merge from onprc19.1 r.65678 to 65757

labkey-bpatel committed:
Fix build issues

labkey-bpatel committed:
Merge from onprc19.1Prod r 65762 to 65949

labkey-bpatel committed:
Remove code duplication from the merge.

labkey-jeckels commented on EHR_ComplianceDB/resources/schemas/dbscripts/sqlserver/EHR_ComplianceDB-12.39-12.40.sql:
```diff
@@ -0,0 +1,2 @@
+ALTER TABLE ehr_compliancedb.Employees ALTER COLUMN Notes NVARCHAR (4000);
```
We should do a matching Postgres script

labkey-bpatel replied:
ah! yes, good catch! now done!

labkey-jeckels commented on ehr/resources/queries/study/treatment_order.query.xml:
```diff
@@ -60,7 +60,7 @@
                         <fk>
                             <fkDbSchema>ehr_lookups</fkDbSchema>
-                            <fkTable>treatment_frequency</fkTable>
+                            <fkTable>Treatment_Frequencies_Active</fkTable>
```
This query looks to be defined via the schema browser for ONPRC. We will need to add it as a custom query in the EHR module to avoid breaking everyone else. Looks like no .query.xml is required, just the SQL:

Select * from ehr_lookups.treatment_Frequency
Where active = 'true'

labkey-bpatel replied:
done

labkey-jeckels commented on ehr/resources/schemas/dbscripts/sqlserver/EHR_Lookups-17.20-17.21.sql:
```diff
@@ -0,0 +1,3 @@
+/* File copied from EHR version upgrade to 17.21 in onprc17.2Prod branch */
+
+ALTER TABLE ehr_lookups.flag_categories ADD datedisabled DATETIME;
```
This file isn't needed anymore. ONPRC is already on 19.1, and bootstrapping servers will run the rollup script.

labkey-bpatel replied:
now removed

labkey-jeckels commented on ehr/resources/schemas/dbscripts/sqlserver/a_ehr_lookups-17.20-17.21.sql:
```diff
@@ -0,0 +1,3 @@
+/* File copied from EHR version upgrade to 17.21 in onprc17.2Prod branch */
+
+ALTER TABLE ehr_lookups.flag_categories ADD datedisabled DATETIME;
```
This file isn't needed anymore. ONPRC is already on 19.1, and bootstrapping servers will run the rollup script.

labkey-bpatel replied:
now removed

labkey-jeckels commented on ehr/resources/schemas/dbscripts/sqlserver/ehr-17.21-17.30.sql:
```diff
@@ -0,0 +1,40 @@
+
+/* ehr-17.20-17.21.sql */
+
+CREATE TABLE ehr.observation_types (
```
This file isn't needed anymore. ONPRC is already on 19.1, and bootstrapping servers will run the rollup script.

labkey-bpatel replied:
removed

labkey-jeckels commented on ehr/resources/schemas/ehr_lookups.xml:
```diff
@@ -2710,6 +2710,9 @@
                     <fkColumnName>entityid</fkColumnName>
                 </fk>
             </column>
+            <column columnName="datedisabled">
```
This is a dupe from lines 2701-2703

labkey-bpatel replied:
removed

labkey-jeckels commented on ehr/resources/schemas/ehr_lookups.xml:
```diff
@@ -2754,6 +2757,9 @@
                     <fkColumnName>entityid</fkColumnName>
                 </fk>
             </column>
+            <column columnName="datedisabled">
```
Looks like a bad merge. This column doesn't exist on this table.

labkey-bpatel replied:
bad merge indeed! removed

labkey-jeckels commented on ehr/resources/views/groupOverlaps.html:
```diff
@@ -154,16 +154,19 @@
                 schemaName: 'study',
                 queryName: queryName,
-                title: 'Group Members'
+                title: 'Group Members Test',
```
This shouldn't include "Test"

labkey-bpatel replied:
done

labkey-jeckels commented on ehr/src/org/labkey/ehr/EHRModule.java:
```diff
@@ -256,6 +256,9 @@ public void moduleStartupComplete(ServletContext servletContext)
         EHRService.get().registerMoreActionsButton(new MarkCompletedButton(this, "study", "parentage", "End Selected Calls", true), "study", "parentage");
         EHRService.get().registerMoreActionsButton(new ExcelImportButton(this, "study", "parentage", "Import Data"), "study", "parentage");
 
+        EHRService.get().registerMoreActionsButton(new MarkCompletedButton(this, "study", "geneticAncestry", "End/Disable Selected", true), "study", "geneticAncestry");
```
These should go into ONPRC_EHRModule instead.

labkey-bpatel replied:
done

labkey-jeckels commented on ehr/src/org/labkey/ehr/EHRServiceImpl.java:
```diff
@@ -456,7 +456,6 @@ public void registerOptionalClinicalHistoryResources(Module module)
         EHRService.get().registerLabworkType(new AntibioticSensitivityLabworkType(module));
         EHRService.get().registerLabworkType(new ChemistryLabworkType(module));
         EHRService.get().registerLabworkType(new HematologyLabworkType(module));
-        EHRService.get().registerLabworkType(new iStatLabworkType(module));
```
To remove this here, we need to add it to WNPRC_EHRModule.doStartupAfterSpringConfig() so they don't lose it.

labkey-bpatel replied:
done

labkey-jeckels commented on ehr/src/org/labkey/ehr/EHRServiceImpl.java:
```diff
@@ -443,7 +443,7 @@ public void registerOptionalClinicalHistoryResources(Module module)
         EHRService.get().registerHistoryDataSource(new DefaultAssignmentEndDataSource(module));
         EHRService.get().registerHistoryDataSource(new DefaultBirthDataSource(module));
-        EHRService.get().registerHistoryDataSource(new DefaultClinicalRemarksDataSource(module));
+       // EHRService.get().registerHistoryDataSource(new DefaultClinicalRemarksDataSource(module));
```
To remove this here, we need to add it to WNPRC_EHRModule.doStartupAfterSpringConfig() so they don't lose it.

Also, let's delete the line instead of commenting it out

labkey-bpatel replied:
done

labkey-jeckels reviewed (changes requested):
Overall code is looking pretty good.

labkey-bpatel committed:
'datedisabled' listed twice - remove duplicate

labkey-bpatel committed:
Remove 17.x scripts (content already part of rolled up scripts). Minor updates to metadata.

labkey-bpatel committed:
Code review changes:
- Add matching postgres script for EHR_ComplianceDB-12.39-12.40.sql
- Add Treatment_Frequencies_Active.sql
- Other minor cleanup

labkey-jeckels commented on EHR_ComplianceDB/resources/schemas/dbscripts/postgresql/EHR_ComplianceDB-12.39-12.40.sql:
```diff
@@ -0,0 +1 @@
+ALTER TABLE ehr_compliancedb.Employees ALTER COLUMN Notes TYPE NVARCHAR (4000);
```
Postgres needs VARCHAR, not NVARCHAR

labkey-bpatel replied:
done

labkey-jeckels reviewed (changes requested):
One Postgres syntax issue, otherwise all set

labkey-bpatel commented on ehr/src/org/labkey/ehr/table/DefaultEHRCustomizer.java:
```diff
@@ -638,18 +615,36 @@ private void customizeStudyData(AbstractTableInfo ti)
+        remarkCol.setLocked(false);
+        remarkCol.setScale(1000000);
+        remarkCol.setLocked(true);
+    }
```
@labkey-jeckels This is throwing below error during study import, since looks like remark column is "locked" and cannot be unlocked/modified this way. For ONPRC's use case, can we just add the scale for 'remark' column in datasets_metadata.xml?

java.lang.IllegalStateException: Can't unlock a ColumnInfo: remark
    at org.labkey.api.data.BaseColumnInfo.setLocked(BaseColumnInfo.java:2179)
    at org.labkey.ehr.table.DefaultEHRCustomizer.expandRemarksColumn(DefaultEHRCustomizer.java:630)
    at org.labkey.ehr.table.DefaultEHRCustomizer.customizeEncountersTable(DefaultEHRCustomizer.java:644)
    at org.labkey.ehr.table.DefaultEHRCustomizer.customizeDataset(DefaultEHRCustomizer.java:463)
    at org.labkey.ehr.table.DefaultEHRCustomizer.customize(DefaultEHRCustomizer.java:123)

labkey-jeckels replied:
In my testing we can just get rid of this completely. The length validation code now allows the length through based on the underlying column in the table.

labkey-bpatel replied:
done

labkey-bpatel committed:
Code review change - use varchar for postgres instead of nvarchar

bbimber commented on ehr/resources/reports/schemas/study/Pedigree/Pedigree.r:
```diff
@@ -13,10 +13,10 @@ library(Rlabkey)
 
 
-if ((length(labkey.data$id) == 0) | (is.na(labkey.data$dam) & is.na(labkey.data$sire))){
+if ((length(labkey.data$id) == 0) || (is.na(labkey.data$dam) & is.na(labkey.data$sire))){
```
the original was probably wrong in multi-animal cases, but I might write this like:

if ((length(labkey.data$id) == 0) || all(is.na(labkey.data$dam)) & is.na(labkey.data$sire)))){
 
which would require all rows to lack dam+sire.

labkey-bpatel replied:
Thanks for the updated snippet. Now updated!

bbimber commented on ehr/resources/queries/study/treatment_order.query.xml:
```diff
@@ -60,7 +60,7 @@
                         <columnTitle>Frequency</columnTitle>
                         <fk>
                             <fkDbSchema>ehr_lookups</fkDbSchema>
-                            <fkTable>treatment_frequency</fkTable>
```
Comment: in general I dont like to encode the FK using a target that is filtered. This will be that legacy data now has broken FKs. In the situation where the desired behavior is to exclude inactive treatment_frequency records, I typically do one of:

1) in the data entry metadata, add this filter. this would mean that any EHR-related entry code would limit new records to the active frequencies only

2) in some cases you can add "<fkDisplayColumnName useRawValue="true"/>", which would at least mean the data region would render a reasonable value. i forget what the raw values of treatment_frequency are, but if they're strings maybe that's viable here.

labkey-bpatel replied:
Good point, change is now reverted with an addition of a filter which will get applied only during insert and update, and not when viewing the data.

bbimber replied:
When did this LK feature get added to support <filters> in the query.xml?   That's useful....

labkey-bpatel replied:
Looks like it got added early last year

bbimber commented on ehr/resources/views/groupOverlaps.html:
```diff
@@ -154,16 +154,19 @@
                 schemaName: 'study',
                 queryName: queryName,
-                title: 'Group Members'
+                title: 'Group Members',
```
why trailing comma

labkey-bpatel replied:
now removed!

bbimber commented on ehr/resources/views/groupOverlaps.html:
```diff
@@ -154,16 +154,19 @@
+            // The Animal group search is working fine now after this change.
             if (groupName)
-                config.removeableFilters = [LABKEY.Filter.create('groupId/name', groupName, LABKEY.Filter.Types.EQUAL)];
+                //config.removeableFilters = [LABKEY.Filter.create('groupId/name', groupName, LABKEY.Filter.Types.EQUAL)];
```
isnt userFilters the new name for removeableFilters?  this change prevents the user from removing the group name filter, doesnt it?

labkey-bpatel replied:
I believe that is the intention given that user enters filter criteria in a form, which then results in a filtered grid in a different panel, having removable filters on this resultant grid would be confusing.

bbimber replied:
I think it's clear that if there was a bug preventing the filter from being applied, that's a problem.

I can see the argument on there being a filter panel above, but by that same logic then shouldnt start/end date also be immutable? It is possible for the top filter fields and DR to get out of sync.  

This is a small enough change that it's probably not a big deal, but in general I would tend toward preserving capabilities for the user. For example, what is they want to do another filter type?  More than one group?  Contains? etc.

bbimber commented on ehr/resources/web/ehr/window/CreateTaskFromIdsWindow.js:
```diff
@@ -234,7 +234,7 @@ Ext4.define('EHR.window.CreateTaskFromIdsWindow', {
         var type = this.down('#type').getValue().type;
         if (type == 'createNew'){
             return {
-                date: this.down('#date').getValue(),
```
dont you want to respect 'date'? otherwise is it going to default to today's date?

labkey-bpatel replied:
There is no 'date' column in ehr.tasks, there is a however a 'duedate' column, so this correction makes sense.

bbimber replied:
OK, i buy that. In reality I would bet that 'date' is often left alone (so defaulting to today on the client-side), and the server-side is going to default duedate to today. I dont think duedate ends up being used for a lot in practice. That all said, this fix is probably the right thing.

bbimber commented on ehr/resources/web/ehr/window/CreateTaskFromRecordsWindow.js:
```diff
@@ -346,7 +346,7 @@ Ext4.define('EHR.window.CreateTaskFromRecordsWindow', {
         EHR.Utils.createTask({
             initialQCState: 'Scheduled',
             existingRecords: existingRecords,
-            taskRecord: {date: date, assignedTo: assignedTo, category: 'task', title: title, formType: this.formType},
```
same question as above. my knowledge of usage here is out of date, but the original scheme was to create the dataset records where 'date' is the date that work is expected to happen. did something change here?

labkey-bpatel replied:
There is no 'date' column in ehr.tasks, there is a however a 'duedate' column, so this correction makes sense.

bbimber replied:
discussion above covers this

labkey-bpatel committed:
Code review change - remove expandRemarksColumn() that resets remarks column scale.

labkey-bpatel committed:
-Replace hardcoded path with moduleProperty in sql queries.
-Remove broken query study.currentBloodAvailable.
-Add module property DCM_NHP_Resources_Container.
-Update automated tests.

labkey-bpatel committed:
Automated test updates

labkey-bpatel committed:
Code review changes

labkey-bpatel committed:
Merge remote-tracking branch 'origin/release20.7-SNAPSHOT' into 20.7_fb_merge_from_onprc19.1

labkey-bpatel committed:
Automated test fix for testArrivalApi(). (Fix for testBloodVolumeApi() wip)

labkey-bpatel committed:
Code review change - remove filtered query as fkTable, instead use a filter that will get applied during insert and update.

labkey-bpatel committed:
Remove testBloodVolumeApi() - this test no longer relevant since LabKey is not managing the calculations & validations, it is now done by a third party tool.

labkey-bpatel committed:
Remove extra ')'. Test fixes.

labkey-bpatel committed:
Add values to Surgical Repairs template. Move skipStudyImportQueryValidation() to the parent class.

labkey-bpatel committed:
Revert 'Add values to Surgical Repair template'.
Use existing & functional utility code to get ehr_lookups.labwork_services store.
Replace 'onprc-snomedtreatmentcombo' with 'ehr-snomedtreatmentcombo' (since the onprc one is not defined).
Some cleanup.

labkey-bpatel committed:
Get LabworkServices store before getting 'servicerequested' (so the store is not loading when servicerequested is selected)

labkey-bpatel committed:
Skip "Billing Validation Notification" - this is broken on the server and not been run successfully by the client.

labkey-bpatel committed:
Test fixes. Associate 'code' column with editorConfig for ehr_lookups.drug_defaults.

labkey-bpatel committed:
Merge remote-tracking branch 'origin/release20.7-SNAPSHOT' into 20.7_fb_merge_from_onprc19.1

labkey-jeckels commented on ehr/test/src/org/labkey/test/tests/onprc_ehr/AbstractGenericONPRC_EHRTest.java:
```diff
@@ -97,6 +97,13 @@ protected EHRClientAPIHelper getApiHelper()
+    @Override
+    protected boolean skipStudyImportQueryValidation()
+    {
+        return true;
```
Is this still needed?

bbimber replied:
Is this for test length / performance reasons?  Is at least one of the ONPRC EHR tests doing a query validation?

labkey-bpatel replied:
This was added temporarily for query validation to pass so i could fix actual test failures but had missed to removed it after. Now removed, and query failures are fixed as well.

labkey-jeckels reviewed (approved):
One question but looks good.

labkey-bpatel committed:
Remove skipStudyImportQueryValidation(), and other related refactor. Query updates (note: these queries still fail the validation, they are just partially updated/fixed).

labkey-jeckels committed:
Clean up queries - delete defunct queries, add missing lists and columns, fix up XML

labkey-jeckels committed:
Fix ONPRC_EHRTest2.testArrivalForm

labkey-jeckels committed:
Fix ONPRC_EHRTest.testNecropsyRequestFlow
